### PR TITLE
[WIP] "fromhost" property's case is lowered when using UDP

### PR DESCRIPTION
**Pull request by nhosoi**
When imudp is used for input, the FROMHOST value is all lowered:
`FROMHOST: 'myclient.example.lab', fromhost-ip: '192.168.122.111', HOSTNAME: 'myCLIENT', ...`
while the case is preserved if it is imtcp:
`FROMHOST: 'myCLIENT.example.lab', fromhost-ip: '192.168.122.111', HOSTNAME: 'myCLIENT', ...`
where the host name is defined as "myCLIENT.example.lab myCLIENT" in /etc/hosts.

These 2 behaviours are preferably matched.

My concern is resolveDNS is a common helper function in runtime/msg.c, which is being called not just from getRcvFrom but multiple places.  The rest may expect all lowercase localName.

### CODE REVIEW

**Comment by davidelang** [comment]
hostnames and DNS names are defined to not be case sensitive, so it should be 
just fine to have things squashed to lower case.

if you have "myCLIENT.example.lab myCLIENT" in /etc/hosts, you can ping myclient 
or MYCLIENT or MyClIeNt and it will work.

why should it matter?

David Lang
k

**Comment by nhosoi** [comment]
We got a bug report which does not clearly state why they need to preserve the cases.

This is a provided example.  They are using a machine generated host name.  There could be some software which requires the case information.
```
/etc/hosts: 10.11.22.33 A0043MtStgRtr01
logfile: May 11 00:00:00 a0043mtstgrtr01 34627: ...
```

**Comment by davidelang** [comment]
On Wed, 13 Jun 2018, Noriko Hosoi wrote:

> We got a bug report which does not clearly state why they need to preserve the cases.
>
> This is a provided example.  They are using a machine generated host name.  There could be some software which requires the case information.
> ```
> /etc/hosts: 10.11.22.33 A0043MtStgRtr01
> logfile: May 11 00:00:00 a0043mtstgrtr01 34627: ...
> ```

please try to find more info. Anything that depends on hostname/dns information 
being case sensitive is going to break in all sorts of ugly ways...

**Comment by nhosoi** [comment]
Thank you, @davidelang.  Understood.
> please try to find more info. 

I'm asking them for more information.

**Comment by rgerhards** [comment]
> These 2 behaviours are preferably matched.
> 
> My concern is resolveDNS is a common helper function in runtime/msg.c, which is being called not just from getRcvFrom but multiple places. The rest may expect all lowercase localName.

It's interesting that actually two different values are returned, depending on module. I would say that's not good and also not intended. Maybe we should file a specific issue for that.

**Comment by rgerhards** [comment]
I am not sure we can merge this PR as is: we need to evaluate implications on existing configs. I guess this at least needs to be configurable.

**Comment by nhosoi** [comment]
Thank you for your comments, @rgerhards.  I agree with you.  Changing the behaviour in resolveDNS should be risky, which may break existing software that depends upon it.

If we are to introduce a new option for preserving the case to imudp, we should do the same to imtcp, as well?  (And to maintain the current behaviour, we should set the different default value to each?  It's getting complexer... :)

**Comment by nhosoi** [comment]
Hi @davidelang, Hi @rgerhards,

I've added an option preservecase to imudp and imtcp module for managing the case of FROMHOST value.  Could you please take a look at the latest patch?  The new patch failed the CI tests, though.  It looks to me the failure did not come from the change I made, but I'm not 100% sure.  If you could give me any advice about the test failure, I'd appreciate it, too.

```
Usage:
module(load="imudp" [preservecase="on"|"off"])
module(load="imtdp" [preservecase="on"|"off"])

If preservecase="on", FROMHOST value is handled in the case sensitive manner.
If preservecase="off", FROMHOST value is handled in the case insensitive manner.

To maintain the current behaviour, the default value of preservecase is "on" for imtcp and "off" for imudp.

See also
https://bugzilla.redhat.com/show_bug.cgi?id=1309698
```

**Comment by rgerhards** [comment]
I am currently working heavily on the CI and will handle issues. If some actually need your attention, I let you know.

Please also create a PR in https://github.com/rsyslog/rsyslog-doc for doc update.

**Comment by rgerhards** [comment]
yeah, you need to do a rebase after I have merged https://github.com/rsyslog/rsyslog/pull/2835 -- I'll let you know. That should solve most of the false positives.

**Comment by rgerhards** [comment]
ok, it was already run through. Please do a rebase when convenient. Note that CI may still fail, probably for a couple of days.

File: plugins/imtcp/imtcp.c [hunk 1]
```diff
@@ -170,7 +172,8 @@ static struct cnfparamdescr modpdescr[] = {
 	{ "keepalive.probes", eCmdHdlrPositiveInt, 0 },
 	{ "keepalive.time", eCmdHdlrPositiveInt, 0 },
 	{ "keepalive.interval", eCmdHdlrPositiveInt, 0 },
-	{ "gnutlsprioritystring", eCmdHdlrString, 0 }
+	{ "gnutlsprioritystring", eCmdHdlrString, 0 },
+	{ "preservecase", eCmdHdlrBinary, 0 },
```
Comment (rgerhards) [inline plugins/imtcp/imtcp.c]: the comma needs to go away ;-)

Comment (nhosoi) [reply plugins/imtcp/imtcp.c]: *Shame shame*  Fixed.  Thanks!

**Comment by nhosoi** [comment]
Thanks, @rgerhards!  I've rebased and pushed the patch.  I'm openining the doc pr, next.

File: runtime/tcpsrv.h [hunk 2]
```diff
@@ -177,6 +178,8 @@ BEGINinterface(tcpsrv) /* name must also be changed in ENDinterface macro! */
 	rsRetVal (*SetbSPFramingFix)(tcpsrv_t*, sbool);
 	/* added v19 -- PascalWithopf, 2017-08-08 */
 	rsRetVal (*SetGnutlsPriorityString)(tcpsrv_t*, uchar*);
+	/* added v?? -- Preserve case in fromhost, 2018-07-?? */
+	rsRetVal (*SetPreserveCase)(tcpsrv_t *pThis, int bPreserveCase);
 ENDinterface(tcpsrv)
 #define tcpsrvCURR_IF_VERSION 20 /* increment whenever you change the interface structure! */
```
Comment (rgerhards) [inline runtime/tcpsrv.h]: while this interface is no longer really enforced, the IF_VERSION should be incremented on interface change. So 21 would be the correct value now.

Comment (nhosoi) [reply runtime/tcpsrv.h]: Thank you, @rgerhards!  I've updated tcpsrv.h and rebased.

File: runtime/tcpsrv.h [hunk 3]
```diff
@@ -177,6 +178,8 @@ BEGINinterface(tcpsrv) /* name must also be changed in ENDinterface macro! */
 	rsRetVal (*SetbSPFramingFix)(tcpsrv_t*, sbool);
 	/* added v19 -- PascalWithopf, 2017-08-08 */
 	rsRetVal (*SetGnutlsPriorityString)(tcpsrv_t*, uchar*);
+	/* added v21 -- Preserve case in fromhost, 2018-08-15 */
+	rsRetVal (*SetPreserveCase)(tcpsrv_t *pThis, int bPreserveCase);
 ENDinterface(tcpsrv)
 #define tcpsrvCURR_IF_VERSION 20 /* increment whenever you change the interface structure! */
```
Comment (rgerhards) [inline runtime/tcpsrv.h]: I guess you overlooked this line ;-) ```#define tcpsrvCURR_IF_VERSION 20```

**Comment by nhosoi** [comment]
Ahhh, sorry about it.  Replaced v21 with v20...

**Comment by rgerhards** [comment]
> Ahhh, sorry about it. Replaced v21 with v20...

mhhh... I  still don't see the 21 in the definition... push forgotten?

**Comment by rgerhards** [comment]
> So sorry... I've pushed a new patch...

not at all a problem sh.. happens ;-) Thx for your persistence, I guess we are now pretty close to merge :-)

**Pull request merged**